### PR TITLE
Integrate `ld+json` structured data for article and news-article

### PR DESCRIPTION
**Pull request by Jeevan-Kishore**
The publisher needs to set `structured_data_news_article` in `theme-attributes` of publisher configuration to enable  `NewsArticle` for each story

`Article` `ld+json` is enabled by default for each type of story

### CODE REVIEW

File: src/structured-data-tags.js [hunk 1]
```diff
@@ -10,11 +10,46 @@ function ldJson(type, fields) {
   };
 }
 
-export function StructuredDataTags({structuredData = {}}, config, pageType, data, {url}) {
+function generateArticleData (story = {}, publisherConfig = {}){
+  const metaKeywords = story.seo && story.seo['meta-keywords'] || [];
+  return {
+    "author": story['author-name'],
+    "editor": story['author-name'],
```
Comment (serendipitist) [inline src/structured-data-tags.js]: could you please check editor and author has same value

File: src/structured-data-tags.js [hunk 2]
```diff
@@ -10,11 +10,51 @@ function ldJson(type, fields) {
...
+function generateArticleData (story = {}, publisherConfig = {}){
+  const metaKeywords = story.seo && story.seo['meta-keywords'] || [];
+  const { authors = [] } = story;
+  return {
+    "author": authors.map(author => ({
+        "@type": "Person",
+        "givenName": author.name,
+        "name": author.name
+    })),
+    "editor": story['owner-name'],
```
Comment (rvgpl) [inline src/structured-data-tags.js]: Do we need `editor` value set? Is it a mandatory field?

File: src/structured-data-tags.js [hunk 3]
```diff
@@ -10,11 +10,51 @@ function ldJson(type, fields) {
...
+    "keywords": metaKeywords,
+    "publisher": publisherConfig['publisher-name'],
```
Comment (rvgpl) [inline src/structured-data-tags.js]: the publisher should be an organization schema object.

File: src/structured-data-tags.js [hunk 4]
```diff
@@ -10,11 +10,51 @@ function ldJson(type, fields) {
...
+    "url": `${publisherConfig['sketches-host']}/${story.slug}`,
+    "dateCreated": new Date(story['created-at']),
+    "dateModified": new Date(story['updated-at']),
+  }
+}
+
+function generateNewsArticleData (story = {}, publisherConfig = {}) {
+  return {
+    'headline' : story.headline,
+    "alternativeHeadline": story.subheadline,
```
Comment (rvgpl) [inline src/structured-data-tags.js]: The alternative headline is not subheadline.

File: src/structured-data-tags.js [hunk 5]
```diff
@@ -10,11 +10,51 @@ function ldJson(type, fields) {
...
+    "image": [`${publisherConfig['cdn-image']}/${story['hero-image-s3-key']}`],
```
Comment (rvgpl) [inline src/structured-data-tags.js]: Are we showing the full res image? Shouldn't we compress and control the dimensions?

File: src/structured-data-tags.js [hunk 6]
```diff
@@ -10,11 +10,51 @@ function ldJson(type, fields) {
...
+    "author": authors.map(author => ({
+        "@type": "Person",
+        "givenName": author.name,
+        "name": author.name
+    })),
+    "editor": story['owner-name'],
+    "keywords": metaKeywords,
+    "publisher": publisherConfig['publisher-name'],
+    "url": `${publisherConfig['sketches-host']}/${story.slug}`,
+    "dateCreated": new Date(story['created-at']),
```
Comment (rvgpl) [inline src/structured-data-tags.js]: `created-at` is same as `published-at`?

File: src/structured-data-tags.js [hunk 7]
```diff
@@ -10,11 +10,51 @@ function ldJson(type, fields) {
...
+    "alternativeHeadline": story.subheadline,
+    "image": [`${publisherConfig['cdn-image']}/${story['hero-image-s3-key']}`],
+    "datePublished": new Date(story['published-at']),
+    "description": story.summary,
+  };
+}
+
+export function StructuredDataTags({structuredData = {}}, config, pageType, data = {}, {url}) {
   const tags = [];
+  const {story = {}} = data.data;
```
Comment (rvgpl) [inline src/structured-data-tags.js]: data.data is weird. :/

File: src/structured-data-tags.js [hunk 8]
```diff
@@ -10,11 +10,60 @@ function ldJson(type, fields) {
...
+
+export function StructuredDataTags({structuredData = {}}, config, pageType, response = {}, {url}) {
   const tags = [];
+  const {story = {}} = response.data;
+  const {config: publisherConfig = {}} = response;
+  const {articleType = ''} = publisherConfig['publisher-settings'] || {};
+
+  const articleData = generateArticleData(story, publisherConfig);
+
+  tags.push(ldJson("Article", articleData));
```
Comment (gja) [inline src/structured-data-tags.js]: inline this variable

File: src/structured-data-tags.js [hunk 9]
```diff
@@ -10,11 +10,60 @@ function ldJson(type, fields) {
...
+    "datePublished": new Date(story['published-at']),
+    "description": story.summary,
+  };
+}
+
+export function StructuredDataTags({structuredData = {}}, config, pageType, response = {}, {url}) {
   const tags = [];
+  const {story = {}} = response.data;
+  const {config: publisherConfig = {}} = response;
+  const {articleType = ''} = publisherConfig['publisher-settings'] || {};
```
Comment (gja) [inline src/structured-data-tags.js]: Super Confusing to read? Is this really worth it?

File: src/structured-data-tags.js [hunk 10]
```diff
@@ -10,11 +10,60 @@ function ldJson(type, fields) {
...
+}
+
+function generateNewsArticleData (story = {}, publisherConfig = {}) {
+  const {alternative = {}} = story.alternative || {};
+  return {
+    'headline' : story.headline,
+    "alternativeHeadline": (alternative.home && alternative.home.default) ? alternative.home.default.headline : "",
+    "image": [`${publisherConfig['cdn-image']}/${story['hero-image-s3-key']}?w=480&auto=format%2Ccompress&fit=max`],
+    "datePublished": new Date(story['published-at']),
+    "description": story.summary,
```
Comment (gja) [inline src/structured-data-tags.js]: description doesn't have to have the full content is it?

File: src/structured-data-tags.js [hunk 11]
```diff
@@ -10,11 +10,60 @@ function ldJson(type, fields) {
...
+      "givenName": author.name,
+      "name": author.name
+    })),
+    "keywords": metaKeywords,
+    "publisher": {
+      "@type": "Organization",
+      "name": publisherConfig['publisher-name'],
+      "logo": {
+        "@type": "ImageObject",
+        "url": themeConfig.logo ? themeConfig.logo : "https://quintype.com/logo.png",
```
Comment (gja) [inline src/structured-data-tags.js]: Do not default this. I think if publisher logo is missing, just do not send anything

Comment (gja) [reply src/structured-data-tags.js]: Also, this is a 404

**Comment by gja** [comment]
Where are the tests?

File: src/structured-data-tags.js (hunk 9 above)
Comment (gja) [reply src/structured-data-tags.js]: Also, what is 'publisher-settings'?

File: src/structured-data-tags.js [hunk 12]
```diff
@@ -10,11 +10,60 @@ function ldJson(type, fields) {
...
+    "image": [`${publisherConfig['cdn-image']}/${story['hero-image-s3-key']}?w=480&auto=format%2Ccompress&fit=max`],
+    "datePublished": new Date(story['published-at']),
+    "description": story.summary,
+  };
+}
+
+export function StructuredDataTags({structuredData = {}}, config, pageType, response = {}, {url}) {
   const tags = [];
+  const {story = {}} = response.data;
+  const {config: publisherConfig = {}} = response;
```
Comment (gja) [inline src/structured-data-tags.js]: why do you need response here?

**Comment by gja** [comment]
Pull Request looks good, just please remove the extra junk in the test

**Pull request merged**